### PR TITLE
Add noderole to dist node, most functions only return primary nodes

### PR DESCRIPTION
**Pull request by lithp**
Wasn't able to add a partial unique index on (groupid, noderole) in order to ensure one primary per group because of [reasons](https://github.com/postgres/postgres/blob/master/src/backend/catalog/indexing.c#L114).

### CODE REVIEW

**Comment by lithp** [comment]
@marcocitus I believe this is ready for you.

**Comment by lithp** [comment]
@marcocitus ping :)

File: src/backend/distributed/citus--6.1-17--6.1-18.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/* citus--6.1-17--6.1-18.sql */
+
+SET search_path = 'pg_catalog';
+
+ALTER TABLE pg_dist_node ADD COLUMN noderole "char" NOT NULL DEFAULT 'p';
```
Comment (marcocitus) [inline src/backend/distributed/citus--6.1-17--6.1-18.sql]: Were you going to add a unique constraint for primaries?

Comment (lithp) [inline src/backend/distributed/citus--6.1-17--6.1-18.sql]: I wanted to, however that's not allowed, [apparently](https://github.com/postgres/postgres/blob/master/src/backend/catalog/indexing.c#L114).

File: src/backend/distributed/citus--6.1-17--6.1-18.sql [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+DROP FUNCTION IF EXISTS master_add_node(text, integer);
+
+CREATE FUNCTION master_add_node(nodename text,
```
Comment (marcocitus) [inline src/backend/distributed/citus--6.1-17--6.1-18.sql]: Would it make sense to add an optional parameter for the noderole in this PR?

File: src/backend/distributed/citus--6.1-17--6.1-18.sql (hunk 1 above)
Comment (marcocitus) [inline src/backend/distributed/citus--6.1-17--6.1-18.sql]: ok, we should probably just verify it when the node is added.

File: src/backend/distributed/Makefile [hunk 3]
```diff
@@ -9,7 +9,7 @@ EXTVERSIONS = 5.0 5.0-1 5.0-2  \
 	5.1-1 5.1-2 5.1-3 5.1-4 5.1-5 5.1-6 5.1-7 5.1-8 \
 	5.2-1 5.2-2 5.2-3 5.2-4 \
 	6.0-1 6.0-2 6.0-3 6.0-4 6.0-5 6.0-6 6.0-7 6.0-8 6.0-9 6.0-10 6.0-11 6.0-12 6.0-13 6.0-14 6.0-15 6.0-16 6.0-17 6.0-18 \
-	6.1-1 6.1-2 6.1-3 6.1-4 6.1-5 6.1-6 6.1-7 6.1-8 6.1-9 6.1-10 6.1-11 6.1-12 6.1-13 6.1-14 6.1-15 6.1-16 6.1-17
+	6.1-1 6.1-2 6.1-3 6.1-4 6.1-5 6.1-6 6.1-7 6.1-8 6.1-9 6.1-10 6.1-11 6.1-12 6.1-13 6.1-14 6.1-15 6.1-16 6.1-17 6.1-18
```
Comment (jasonmp85) [inline src/backend/distributed/Makefile]: :+1:

**Comment by sumedhpathak** [comment]
@lithp @marcocitus is this PR now defunct? Or do we need it for auto-ha?

**Comment by marcocitus** [comment]
I think it's not a must-have for auto-failover, only for the ability to read from secondaries. However, it would be harder to change it later on, so therefore a should have.

The must-haves for auto-failover is to provide a way to update the primary nodename in pg_dist_node and pg_dist_shard_placement, and a way to block writes to a node.

**Comment by lithp** [comment]
Some notes:
- It's probably better for `noderole` to be an enum rather than a char. This is more difficult for the user but much better for the user.
- Andres brought up a good idea, maybe it makes more sense for each `shard_placement` tuple to spawn multiple `ShardPlacement` structs, one per node in the group.

**Comment by sumedhpathak** [comment]
@lithp Can we close this PR then? Thanks!

**Comment by marcocitus** [comment]
Superseded by #1476

**Pull request closed without merging**